### PR TITLE
Export and import content revisions

### PR DESCRIPTION
**Pull request by pbauer**
#105

### CODE REVIEW

File: src/collective/exportimport/import_content.py [hunk 1]
```diff
@@ -399,6 +407,120 @@ def import_new_content(self, data):  # noqa: C901
...
+        return new
+
+    def save_revision(self, obj, item, initial=False):
+        """Save revision manually to set dates and changenote from exported data."""
+        rt = api.portal.get_tool("portal_repository")
+
+        modified = dateutil.parser.parse(item["modified"])
+        # add one millisecond to prevent created being before first revision
+        modified  = modified + timedelta(milliseconds=1)
+        timestamp = datetime.timestamp(modified)
```
Comment (radiusJS) [inline src/collective/exportimport/import_content.py:511]: `datetime.timestamp` is only available from python 3.3 up. For python 2 we could use `time.mktime.`

```suggestion
         try:
            timestamp = datetime.timestamp(modified)
        except AttributeError:
            # support python 2.7 
            import time
            timestamp = time.mktime(modified.timetuple())
```

Comment (pbauer) [reply src/collective/exportimport/import_content.py]: fixed

**Comment by pbauer** [comment]
I think this is ready.

File: src/collective/exportimport/import_content.py [hunk 2]
```diff
@@ -399,6 +409,140 @@ def import_new_content(self, data):  # noqa: C901
...
+        """
+        portal_workflow = api.portal.get_tool("portal_workflow")
+
+        # Disable automatic versioning!
+        portal_types = api.portal.get_tool("portal_types")
+        fti = portal_types.get(item["@type"])
+        repo_tool = api.portal.get_tool("portal_repository")
+        policy = None
+        policies = repo_tool._version_policy_mapping.get(item["@type"], [])
+        versioning_behavior = None
```
Comment (mauritsvanrees) [inline src/collective/exportimport/import_content.py:425]: This code part could do with a small explanation: "We are temporarily disabling versioning and will re-enable it at the end.".

**Review by mauritsvanrees — APPROVED** [review]
LGTM, and it works for me when exporting a few Dexterity items from Plone 4.3 into Plone 6.0. The history viewlet looks the same afterwards. So fine to merge.

Just two inline comments that could be addressed. But since it works, I approve.
Thanks!

File: src/collective/exportimport/import_content.py [hunk 3]
```diff
@@ -338,6 +342,12 @@ def import_new_content(self, data):  # noqa: C901
                         )
                     )
 
+            if self.import_old_revisions and item.get("exportimport.versions"):
+                new = self.import_versions(container, item)
```
[the remark this answers is not in the document]
Comment (pbauer) [reply src/collective/exportimport/import_content.py:346]: Good point. I'll try to refactor the logic into a method `import_item` and get rid of some duplication

[the remark this answers is not in the document]
Comment (pbauer) [reply src/collective/exportimport/import_content.py:346]: That might have to wait until the next version though...

**Comment by fredvd** [comment]
I have 'end user' tested with a small Plone 4.3.20 site with around 90 content items of which 10 documents.  The revisions are exported and imported, but when I look into the revision details, only the number of revisions match.  Import into Plone 6.0a4 classic setup.

* The date of all revisions is set to the date of the newest revision(live version). 
* Also If I try to compare revisions no changes at all displayed. When I do this in the Plone 4 site I do see changes in the main text being highlighted etc. etc. 
* the creation, initial revision and publish lines in the history have the correct author set.  All subsequent revisions are ownd by the user under which I ran the import (admin).

**Comment by pbauer** [comment]
Oh, yes, I totally forgot a very important detail! I guess you tested using Archetypes content oin 4.3.x?
To properly serialize versions of AT content you need this fix in plone.restapi: https://github.com/plone/plone.restapi/issues/1335). It is merged in the branch 7.x.x of plone.restapi (8.x.x no longer supports AT) and until a new releasae of 7.x is done you need a source checkout. Sorry.

**Pull request merged**